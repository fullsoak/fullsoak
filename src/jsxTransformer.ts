import { LogDebug, readFileToString } from "./utils.ts";

type GenericJsTransform = (
  code: string,
  // deno-lint-ignore no-explicit-any
  options: any,
) => Promise<{ code: string }>;

let _fallbackTransform: GenericJsTransform;

const getTscJsTransform = async (): Promise<GenericJsTransform> => {
  if (_fallbackTransform) return _fallbackTransform;
  const { transpileModule } = await import("typescript");
  const retVal: GenericJsTransform = (code, options) => {
    const res = transpileModule(code, {
      compilerOptions: {
        jsxImportSource: "preact",
        jsx: "react-jsx",
        jsxFactory: "h",
        jsxFragmentFactory: "Fragment",
        target: "es6",
        module: "esnext",
      },
      ...options,
    });
    // @NOTE there exist other fields returned by `typescript::transpileModule`
    return Promise.resolve({ code: res.outputText });
  };
  return _fallbackTransform = retVal;
};

let _transformFileFn: GenericJsTransform;
let _transformFn: GenericJsTransform;
let _minifyFn: GenericJsTransform;

/**
 * retrieve (and cache) the JS transpiling functions by asynchronously
 * importing the native binding functions, with fallback
 * results if/when a native binding function fails to load;
 * e.g: https://github.com/denoland/deploy_feedback/issues/802#issuecomment-2661089887
 */
export const getJsTransformFns: () => Promise<{
  transformFile: GenericJsTransform;
  transform: GenericJsTransform;
  minify: GenericJsTransform;
}> = async () => {
  if (_transformFileFn && _transformFn && _minifyFn) {
    return {
      transformFile: _transformFileFn,
      transform: _transformFn,
      minify: _minifyFn,
    };
  }

  try {
    const swcCore = await import("@swc/core");
    _transformFileFn = swcCore.transformFile;
    _transformFn = swcCore.transform;
    _minifyFn = swcCore.minify;
  } catch (e) {
    LogDebug("unable to load @swc/core", (e as Error).stack, "falling back...");

    _transformFn = async (code: string, opts: unknown) => {
      const tscTx = await getTscJsTransform();
      return tscTx(code, opts);
    };

    _transformFileFn = async (filePath, opts) =>
      _transformFn(
        await readFileToString(filePath),
        opts,
      );

    let _doMinify = (x: string) => ({ code: x }); // @TODO find a better fallback
    try {
      _doMinify = (await import("uglify-js")).minify || _doMinify;
    } catch (_) {
      // ignored
    }
    _minifyFn = (src) => Promise.resolve(_doMinify(src));

    LogDebug("successfully registered fallback tools for @swc/core");
  }

  return {
    transformFile: _transformFileFn,
    transform: _transformFn,
    minify: _minifyFn,
  };
};
