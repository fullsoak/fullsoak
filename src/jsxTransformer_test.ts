import { assertSnapshot } from "@std/testing/snapshot";
import { _internal } from "./jsxTransformer.ts";

const { getTscJsTransform } = _internal;

const sampleJsx = `
  import { Foo } from "./Foo/index.tsx";
  import {Bar}from"../Bar.tsx";import { Chaz}  from   "Chaz.tsx"
  import * as Daz from"./Daz/file.ts";
  import * as Daz1 from     "./Daz1/file.ts";
  export const Eas = () => {
    return <>{Daz.getSomething()}
    <Foo />
      <Bar foo="bar" chaz={<Chaz onClick={() => console.log("Chaz clicked")}/>}></Bar>
      <Daz1 />
    </>;
  }
`;

Deno.test("tsc transformation", async (t) => {
  const tx = await getTscJsTransform();
  const { code } = await tx(sampleJsx, {});
  await assertSnapshot(t, code);
});
