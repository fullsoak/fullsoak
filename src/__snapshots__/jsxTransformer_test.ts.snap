export const snapshot = {};

snapshot[`tsc transformation 1`] = `
'import { jsx as _jsx, Fragment as _Fragment, jsxs as _jsxs } from "preact/jsx-runtime";
import { Foo } from "./Foo/index.tsx";
import { Bar } from "../Bar.tsx";
import { Chaz } from "Chaz.tsx";
import * as Daz from "./Daz/file.ts";
import * as Daz1 from "./Daz1/file.ts";
export const Eas = () => {
    return _jsxs(_Fragment, { children: [Daz.getSomething(), _jsx(Foo, {}), _jsx(Bar, { foo: "bar", chaz: _jsx(Chaz, { onClick: () => console.log("Chaz clicked") }) }), _jsx(Daz1, {})] });
};
'
`;
